src/main/java/org/opensearch/neuralsearch/query/AgenticSearchQueryBuilder.java

```diff
@@ -0,0 +1,266 @@
+     *    "agent_id": "string",
+     *    "query_fields": ["string", "string"..]
+     *    }
+     * }
+     *
+     */
+
+    public static AgenticSearchQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        AgenticSearchQueryBuilder agenticSearchQueryBuilder = new AgenticSearchQueryBuilder();
+
+        String currentFieldName = null;
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if (QUERY_TEXT_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                    agenticSearchQueryBuilder.queryText = parser.text();
+                } else if (AGENT_ID.match(currentFieldName, parser.getDeprecationHandler())) {
+                    agenticSearchQueryBuilder.agentId = parser.text();
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(), "Unknown field [" + currentFieldName + "]");
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                if (QUERY_FIELDS.match(currentFieldName, parser.getDeprecationHandler())) {
+                    List<String> fieldsList = new ArrayList<>();
+                    while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                        fieldsList.add(parser.text());
+                    }
+                    agenticSearchQueryBuilder.queryFields = fieldsList;
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(), "Unknown field [" + currentFieldName + "]");
+                }
+            }
+        }
+
+        // Validate mandatory fields
+        if (agenticSearchQueryBuilder.queryText == null || agenticSearchQueryBuilder.queryText.trim().isEmpty()) {
+            throw new ParsingException(parser.getTokenLocation(), "[" + QUERY_TEXT_FIELD.getPreferredName() + "] is required");
+        }
+        if (agenticSearchQueryBuilder.agentId == null || agenticSearchQueryBuilder.agentId.trim().isEmpty()) {
+            throw new ParsingException(parser.getTokenLocation(), "[" + AGENT_ID.getPreferredName() + "] is required");
+        }
+
+        return agenticSearchQueryBuilder;
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        // Idempotent check - if already rewritten, return the cached result
+        if (rewrittenQuery != null) {
+            return rewrittenQuery;
+        }
+
+        if (ML_CLIENT == null) {
+            throw new IllegalStateException("ML client not initialized");
+        }
+
+        // the queryRewrite is expected at QueryCoordinator level
+        if (!(queryRewriteContext instanceof QueryCoordinatorContext)) {
+            throw new IllegalStateException(
```

Comment:
this exception is properly never thrown if this query builder is actually working